src/storage/workspace_storage.rs

```diff
@@ -311,13 +356,13 @@ impl WorkspaceStorage {
             )
             .await
             .map_err(fs_to_python_error)
-        })
+        }))
     }
 
-    fn get_need_sync_entries(&self) -> FutureIntoCoroutine {
-        let ws = self.0.clone();
+    fn get_need_sync_entries(&self) -> PyResult<FutureIntoCoroutine> {
+        let ws = self.get_storage()?;
 
-        FutureIntoCoroutine::from(async move {
+        Ok(FutureIntoCoroutine::from(async move {
             ws.get_need_sync_entries()
```

Comment:
Why don't you call `get_storage` inside the coroutine, the scope seems the same for your guard and the second `Arc` seems not needed ? :thinking:

Reply:
1. Doing that require to clone `Self` is it something better ?
2. That way it prevent to send a job to `tokio` if the workspace storage is dropped, but this case should not append (that mean we use the `WorkspaceStorage` outside the context manager)

Reply:
You clone self in `get_storage`

Reply:
On another thought, this isn't possible because (to remove the second `Arc`):
1. I don't (and clippy) want to hold a `MutexGuard` between await point.
2. `WorkspaceStorage` (pure rust) isn't clonable and I don't want to make it so because it have to manage a connection database and we better have only 1 instance managing that behind a `Arc`.